### ISSUE TITLE
SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 35396

### ISSUE THREAD

**Issue by dhruv048** [adamlofts/mysql1_dart#60]
Tried sql for the first time but i am getting this error
SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 35396

my code:
 var settings = new ConnectionSettings(
        host: 'localhost',
        port: 3306,
        user: 'flutter_admin',
        password: 'flutter_admin_password',
        db: 'flutter_admin'
    );
    final conn =await MySqlConnection.connect(settings);

    var result = await conn.query(
        'insert into location (id, live_location) values (?, ?)',
        ['Bob', 25]);
    print('Inserted row id=${result.insertId}');

please help me, Thanks

**Comment by julius-stuemmler** [comment]
Hey guys!

I have the same error like @dhruv048.
I found a different behaviour between simple dart-code (Package integrated in Console-App.) and flutter-code (Package integrated in Mobile-Application.). Also interesting: the port is incremented by 2 after retrying the connection.

ERROR-MESSAGE:
`E/flutter (12763): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 43248
E/flutter (12763): 
E/flutter (12763): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 43250
E/flutter (12763): 
E/flutter (12763): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 43252
E/flutter (12763): 
D/DecorView(12763): onWindowFocusChangedFromViewRoot hasFocus: true, DecorView@c0e0f2f[MainActivity]
Lost connection to device.
`

Code is simplified.

FLUTTER-CODE-EXAMPLE (android: 10.0.15):
`import 'package:flutter/material.dart';
import 'package:mysql1/mysql1.dart';

void main() => runApp(new MyApp());

String username='';

class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter App with MYSQL',
      home: new MyHomePage(),

    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  TextEditingController name = new TextEditingController();
  TextEditingController email = new TextEditingController();
  TextEditingController mobile = new TextEditingController();

  List data = [];

  var conn;

  void connect() async{
    conn = await MySqlConnection.connect(ConnectionSettings(host: 'localhost', user: 'root', db: 'registereduser'));
  }

  void printData() async{
    var result = await conn.query("SELECT * FROM `registereduser`");
    print(result);
  }

  @override
  void initState() {
    connect();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text("Register"),),
      body: Container(
        child: Center(
          child: Column(
            children: <Widget>[
              Text("Username",style: TextStyle(fontSize: 18.0),),
              TextField(
                controller: name,
                decoration: InputDecoration(
                    hintText: 'name'
                ),
              ),
              Text("Email",style: TextStyle(fontSize: 18.0),),
              TextField(
                controller: email,
                decoration: InputDecoration(
                    hintText: 'Email'
                ),
              ),
              Text("Mobile",style: TextStyle(fontSize: 18.0),),
              TextField(
                controller: mobile,
                decoration: InputDecoration(
                    hintText: 'Mobile'
                ),
              ),
              RaisedButton(
                child: Text("Register"),
                onPressed: (){
                },
              ),
              Container(height: 20,),
              RaisedButton(
                child: Text("Load data"),
                onPressed: (){
                  connect();
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}`

DART-EXAMPLE:
`import 'package:cli/cli.dart' as cli;
import 'package:mysql1/mysql1.dart';

Future main() async {
  // Open a connection (testdb should already exist)
  final conn = await MySqlConnection.connect(ConnectionSettings(
      host: 'localhost', port: 3306, user: 'root', db: 'test'));

  // Query the database using a parameterized query
  var results = await conn
      .query('select * from user');
  for (var row in results) {
    print(row);
  }

  // Finally, close the connection
  await conn.close();
}
`

I would be really thankful, if you can fix this error. The package is actually really fine.

**Comment by DansMatee** [comment]
I found a solution on Stack Overflow that helped me with this problem. By replacing `host: 'localhost'` with `host: '10.0.2.2'` It seemed to work perfectly.

**Comment by julius-stuemmler** [comment]
@DansMatee 

Yeah possible solution, but I'm still getting this error using a IP of a server, where mysql is installed. Do you have a solution or idea for that?

**Comment by ItzcoR** [comment]
replacing the host for "10.0.2.2" did the trick for me, another detail is that if you are getting a conection refused, despite you having no password just comment the password in ConnectionSettings

**Comment by Adil-Rangila** [comment]
> I found a solution on Stack Overflow that helped me with this problem. By replacing `host: 'localhost'` with `host: '10.0.2.2'` It seemed to work perfectly.

thanks your solution is worked for me

**Comment by flutterq** [comment]
All Possible solution added [[Solved] SocketException: OS Error: Connection refused, errno = 111 in flutter](https://fluttercorner.com/socketexception-os-error-connection-refused-errno-111-in-flutter/)

**Comment by yogithesymbian** [comment]
in my case , instead of local address , i use ethernet ip address ( wireless ) , my android emulator cant ping my local macbook.

**Comment by bmpsst511** [comment]
In case if you use Django as your backend, you need to run with `python manage.py runserver your_ip:port`. And replace your localhost IP address to Django ip address in flutter project if you are trying to run on an iPhone device.
